bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/provider/ScriptModuleTypeProvider.java

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.automation.module.script.internal.provider;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameter;
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameter.Type;
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameterBuilder;
+import org.eclipse.smarthome.config.core.ParameterOption;
+import org.eclipse.smarthome.core.common.registry.ProviderChangeListener;
+import org.openhab.core.automation.Visibility;
+import org.openhab.core.automation.module.script.ScriptEngineFactory;
+import org.openhab.core.automation.module.script.internal.handler.AbstractScriptModuleHandler;
+import org.openhab.core.automation.module.script.internal.handler.ScriptActionHandler;
+import org.openhab.core.automation.module.script.internal.handler.ScriptConditionHandler;
+import org.openhab.core.automation.type.ActionType;
+import org.openhab.core.automation.type.ConditionType;
+import org.openhab.core.automation.type.ModuleType;
+import org.openhab.core.automation.type.ModuleTypeProvider;
+import org.openhab.core.automation.type.Output;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class dynamically provides ScriptActionType and ScriptConditionType {@link ModuleType}s. This class is necessary
+ * because there is no other way to provide dynamic {@link ParameterOption}s for {@link ModuleType}s.
+ *
+ * @author Scott Rushworth - Initial contribution
+ */
+@Component(immediate = true)
```

Comment:
I assume there is a consumer of the `ModuleTypeProvider` components and if there is no one this service does not bring any value.

So, do you see any reason we need to use immediate here?
I assume it can (and if it can it should) be removed.

Reply:
See also https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/provider/file/ModuleTypeFileProvider.java#L30 as another example

Reply:
@davidgraeff I'm not sure what this example is supposed to show.

Reply:
> I assume there is a consumer of the ModuleTypeProvider components 

I've seached for one, but haven't found it. But when the component annotation is removed, the ModuleTypes do not show up in the REST API.

> So, do you see any reason we need to use immediate here?

For this file, I followed most of the patterns in [MediaActionTypeProvider](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.automation.module.media/src/main/java/org/openhab/core/automation/module/media/internal/MediaActionTypeProvider.java#L48), which uses immediate. Same for [ScriptedCustomModuleTypeProvider](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/internal/ScriptedCustomModuleTypeProvider.java#L33). The ModuleTypes will also need to be available before any rules are loaded that require them.

However, on my test system, I did not see any issues removing immediate. I'll try it tonight on my production system.

Reply:
With immediate removed, my production system did not have an issue. I will remove this.